### PR TITLE
3067 restricted user has place and contact

### PR DESCRIPTION
**Pull request by estellecomment**
#3067

### CODE REVIEW

File: tests/karma/unit/controllers/edit-user.js [hunk 1]
```diff
@@ -271,12 +271,93 @@ describe('EditUserCtrl controller', function() {
...
+      // when
+      scope.editUser();
+
+      // expect
+      chai.expect(scope.errors).not.to.have.property('facility_id');
+      chai.expect(scope.errors).not.to.have.property('contact_id');
+    });
+
+
+    var jQuery;
```
Comment (alxndrsn) [inline tests/karma/unit/controllers/edit-user.js]: Could this be stubbed/mocked/spied, and then reset, in the standard sinon way?  Also might be neater including this declaration with all others at the top of the outer `define` function.

Comment (alxndrsn) [inline tests/karma/unit/controllers/edit-user.js]: Could this be stubbed/mocked/spied, and then reset, in the standard sinon way?  Also might be neater including this declaration with all others at the top of the outer `define` function.

File: static/js/controllers/edit-user.js [hunk 2]
```diff
@@ -133,12 +143,18 @@
         if (!updatingSelf) {
           // users don't have permission to update their own security settings
           settings.roles = getRoles($scope.editUserModel.type, true);
```
Comment (alxndrsn) [inline static/js/controllers/edit-user.js]: Is this code duplicated in `computeFields()` deliberately?

File: static/js/controllers/edit-user.js [hunk 3]
```diff
@@ -101,13 +108,16 @@
...
       var validateName = function() {
-        if (!$scope.editUserModel.name) {
-          $scope.errors.name = $translate.instant('field is required', {
-            field: $translate.instant('User Name')
-          });
-          return false;
+        return validateRequired('name', 'User Name');
+      };
+
+      var validateRestrictedUser = function() {
```
Comment (alxndrsn) [inline static/js/controllers/edit-user.js]: I'm slightly uncomfortable that `validateRestrictedUser()` returns `true` if the user is _not_ restricted.

File: static/js/controllers/edit-user.js [hunk 4]
```diff
@@ -133,12 +143,18 @@
...
           settings.roles = getRoles($scope.editUserModel.type, true);
-          settings.facility_id = $('#edit-user-profile [name=facility]').val();
-          settings.contact_id = $('#edit-user-profile [name=contact]').val();
+          settings.facility_id = $scope.editUserModel.facility_id;
+          settings.contact_id = $scope.editUserModel.contact_id;
         }
         return settings;
       };
 
+      var computeFields = function() {
```
Comment (alxndrsn) [inline static/js/controllers/edit-user.js]: I don't understand why these fields aren't bound in the standard angular way.

File: static/js/controllers/edit-user.js [hunk 5]
```diff
@@ -133,12 +143,18 @@
...
+        $scope.editUserModel.roles = getRoles($scope.editUserModel.type, true);
```
Comment (alxndrsn) [inline static/js/controllers/edit-user.js]: Does this allow users to change their own roles?  Having this here would seem to contradict the comment in `getSettingsUpdates()`:

>     // users don't have permission to update their own security settings

File: tests/karma/unit/controllers/edit-user.js [hunk 6]
```diff
@@ -271,12 +271,93 @@ describe('EditUserCtrl controller', function() {
...
+    // If you use this, don't forget to reset: you're mocking out
+    // jQuery for all unit tests.
+    var mockOutjQueryFormFields = function() {
+      jQuery = window.$;
+      window.$ = sinon.stub();
+      window.$.withArgs('#edit-user-profile [name=contact]').returns(
+        {val: function(){ return model.contact_id; }});
```
Comment (alxndrsn) [inline tests/karma/unit/controllers/edit-user.js]: Tying jQuery to the angular model like this seems odd.  I've tried to find other karma tests that we've written which mock jquery at all but haven't been successful.

**Comment by alxndrsn** [comment]
There's still a few queries unanswered from my last review.

File: tests/karma/unit/controllers/edit-user.js (hunk 6 above)
Comment (estellecomment) [inline tests/karma/unit/controllers/edit-user.js]: Filed : Non-angular select 2 fields : #3444

File: tests/karma/unit/controllers/edit-user.js (hunk 1 above)
Comment (estellecomment) [inline tests/karma/unit/controllers/edit-user.js]: done

Comment (estellecomment) [inline tests/karma/unit/controllers/edit-user.js]: done

File: static/js/controllers/edit-user.js (hunk 3 above)
Comment (estellecomment) [inline static/js/controllers/edit-user.js]: renamed

File: static/js/controllers/edit-user.js (hunk 2 above)
Comment (estellecomment) [inline static/js/controllers/edit-user.js]: no, fixed

File: static/js/controllers/edit-user.js (hunk 4 above)
Comment (estellecomment) [inline static/js/controllers/edit-user.js]: Select2 fields don't work happily with the model without extra work. 
Filed : Non-angular select 2 fields : #3444

File: tests/karma/unit/controllers/edit-user.js [hunk 7]
```diff
@@ -271,12 +271,93 @@ describe('EditUserCtrl controller', function() {
       chai.expect(scope.errors).not.to.have.property('password');
     });
 
+    it('must have associated place if user type is restricted user', function() {
+      UpdateUser.returns(KarmaUtils.mockPromise());
+      dbQuery.returns(KarmaUtils.mockPromise());
+      createController();
+
+      scope.editUserModel.type = 'district-manager';
+      scope.editUserModel.facility_id = undefined;
```
Comment (alxndrsn) [inline tests/karma/unit/controllers/edit-user.js]: >  So `null` it is!

Still says undefined :thinking:

File: static/js/controllers/edit-user.js [hunk 8]
```diff
@@ -82,34 +88,59 @@
...
+        }
+
+        // if existing user : needs both fields, or none
+        if ($scope.editUserModel.password) {
+          return validatePasswordAndConfirm();
         }
-        if ($scope.editUserModel.password &&
-            $scope.editUserModel.password !== $scope.editUserModel.passwordConfirm) {
+        if ($scope.editUserModel.passwordConfirm) {
           $scope.errors.password = $translate.instant('Passwords must match');
```
Comment (alxndrsn) [inline static/js/controllers/edit-user.js]: I think we already returned above if `$scope.editUserModel.password` is set, so this may be dead code.

Comment (alxndrsn) [inline static/js/controllers/edit-user.js]: ...or perhaps not, but why not just call `validatePasswordAndConfirm()` in any case?  It looks like you could replace 108-115 with:
```
if ($scope.editUserModel.password || $scope.editUserModel.password) {
  return validatePasswordAndConfirm();
}
return true;

File: static/js/controllers/edit-user.js [hunk 9]
```diff
@@ -82,34 +88,59 @@
...
-          });
+      var validateConfirmPasswordMatches = function() {
+        if ($scope.editUserModel.password !== $scope.editUserModel.passwordConfirm) {
+          $scope.errors.password = $translate.instant('Passwords must match');
           return false;
         }
         return true;
       };
 
+      var validatePasswordAndConfirm = function() {
```
Comment (alxndrsn) [inline static/js/controllers/edit-user.js]: `validatePasswordFields()` clearer?

File: templates/modals/edit_user.html [hunk 10]
```diff
@@ -52,19 +52,21 @@
       </select>
     </div>
 
-    <div class="form-group" mm-auth="can_configure">
+    <div class="form-group" mm-auth="can_configure" ng-class="{'has-error': errors.facility_id}">
       <label for="facility" translate>Facility</label>
       <select id="facility" name="facility">
         <option ng-value="editUserModel.facility" ng-if="editUserModel.facility"></option>
       </select>
+      <span class="help-block">{{errors.facility_id}}</span>
```
Comment (alxndrsn) [inline templates/modals/edit_user.html]: Do these help blocks need hiding explicitly if there's no error?

File: static/js/controllers/edit-user.js (hunk 8 above)
Comment (estellecomment) [inline static/js/controllers/edit-user.js]: Ha I couldn't find something I was happy with, this looks good, thanks!

File: templates/modals/edit_user.html (hunk 10 above)
Comment (estellecomment) [inline templates/modals/edit_user.html]: nope.

**Pull request merged**